documentation/examples/basic/akkahttp-server/src/main/scala/sample/Api.scala

```diff
@@ -18,10 +22,13 @@ object Api extends ApiAlg with Endpoints with JsonEntitiesFromCodecs with BasicA
       } ~
       maybe.implementedBy { _ =>
         if (util.Random.nextBoolean()) Some(()) else None
-      } ~ auth.implementedBy { credentials =>
-        println(s"Authenticated request: ${credentials.username}")
-        if (Random.nextBoolean()) Some(())
-        else None // Randomly return a forbidden
+      } ~ auth.implementedBy {
+        case Some(credentials) =>
+          println(s"Authenticated request: ${credentials.username}")
+          if (Random.nextBoolean()) Some(())
+          else None // Randomly return an unauthorized
+        case None =>
+          None // Missing credentials. always return an unauthorized
```

Comment:
I'm not certain that the `Forbidden` status code means "invalid credentials". From experience, I think most servers would return a `401 Unauthorized` status code if the client provides invalid credentials.

From https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403:

> The HTTP 403 Forbidden client error status response code indicates that the server understood the request but refuses to authorize it.
> 
> This status is similar to 401, but in this case, re-authenticating will make no difference. The access is permanently forbidden and tied to the application logic, such as insufficient rights to a resource.

So I think `Forbidden` reflects a lack of permission to perform the action. The user is properly authenticated but doesn't have the right to perform the action.

Reply:
Yes, that’s how it works in `BasicAuthentication`: when the request is processed by the server interpreter, it is rejected with an `Unauthorized` response in case the credentials are missing. Then, the endpoint implementation can signal that the authenticated user lacked the right to access the resource by returning `None` (in which case the server produces a `Forbidden` response).